google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimer.java

```diff
@@ -0,0 +1,168 @@
+import io.grpc.ManagedChannel;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+import org.threeten.bp.Duration;
+
+/**
+ * A channel warmer that ensures that a Bigtable channel is ready to be used before being added to
+ * the active {@link com.google.api.gax.grpc.ChannelPool}.
+ *
+ * <p>This implementation is subject to change in the future, but currently it will prime the
+ * channel by sending a ReadRow request for a hardcoded, non-existent row key.
+ */
+@BetaApi("Channel priming is not currently stable and might change in the future")
+class BigtableChannelPrimer implements ChannelPrimer {
+  private static Logger LOG = Logger.getLogger(BigtableChannelPrimer.class.toString());
+
+  static ByteString PRIMING_ROW_KEY = ByteString.copyFromUtf8("nonexistent-priming-row");
+  private static Duration PRIME_REQUEST_TIMEOUT = Duration.ofSeconds(30);
+
+  private final EnhancedBigtableStubSettings settingsTemplate;
+  private final List<String> tableIds;
+
+  static BigtableChannelPrimer create(
+      Credentials credentials,
+      String projectId,
+      String instanceId,
+      String appProfileId,
+      List<String> tableIds) {
+    EnhancedBigtableStubSettings.Builder builder =
+        EnhancedBigtableStubSettings.newBuilder()
+            .setProjectId(projectId)
+            .setInstanceId(instanceId)
+            .setAppProfileId(appProfileId)
+            .setCredentialsProvider(FixedCredentialsProvider.create(credentials))
+            .setExecutorProvider(
+                InstantiatingExecutorProvider.newBuilder().setExecutorThreadCount(1).build());
+
+    // Disable retries for priming request
+    builder
+        .readRowSettings()
+        .setRetrySettings(
+            builder
+                .readRowSettings()
+                .getRetrySettings()
+                .toBuilder()
+                .setMaxAttempts(1)
+                .setJittered(false)
+                .setInitialRpcTimeout(PRIME_REQUEST_TIMEOUT)
+                .setMaxRpcTimeout(PRIME_REQUEST_TIMEOUT)
+                .setTotalTimeout(PRIME_REQUEST_TIMEOUT)
+                .build());
+    return new BigtableChannelPrimer(builder.build(), tableIds);
+  }
+
+  private BigtableChannelPrimer(
+      EnhancedBigtableStubSettings settingsTemplate, List<String> tableIds) {
+    Preconditions.checkNotNull(settingsTemplate, "settingsTemplate can't be null");
+    this.settingsTemplate = settingsTemplate;
+    this.tableIds = ImmutableList.copyOf(tableIds);
+  }
+
+  @Override
+  public void primeChannel(ManagedChannel managedChannel) {
+    try {
+      primeChannelUnsafe(managedChannel);
+    } catch (IOException | RuntimeException e) {
+      LOG.warning(
+          String.format("Unexpected error while trying to prime a channel: %s", e.getMessage()));
+    }
```

Comment:
(see codecov warning) is this really not covered?

Reply:
That branch is for unexpected errors, I added a test to handle that as well